cartography/intel/gcp/bigquery.py

```diff
@@ -0,0 +1,332 @@
+
+    :type bigquery: The GCP Bigquery resource object
+    :param bigquery: The Bigquery  resource object created by googleapiclient.discovery.build()
+
+    :type project_id: str
+    :param project_id: Current Google Project Id
+
+    :rtype: list
+    :return: List of Bigquey Datasets
+    """
+    try:
+        datasets = []
+        request = bigquery.datasets().list(project=project_id, maxResults=5000)
+        while request is not None:
+            response = request.execute()
+            if 'datasets' in response:
+                datasets.extend(response.get('datasets', []))
+            request = bigquery.datasets().list_next(previous_request=request, previous_response=response)
+        if common_job_parameters.get('pagination', {}).get('bigquery', None):
+            pageNo = common_job_parameters.get("pagination", {}).get("bigquery", None)["pageNo"]
+            pageSize = common_job_parameters.get("pagination", {}).get("bigquery", None)["pageSize"]
+            totalPages = len(datasets) / pageSize
+            if int(totalPages) != totalPages:
+                totalPages = totalPages + 1
+            totalPages = int(totalPages)
+            if pageNo < totalPages or pageNo == totalPages:
+                logger.info(f'pages process for datasets {pageNo}/{totalPages} pageSize is {pageSize}')
+            page_start = (
+                common_job_parameters.get('pagination', {}).get('bigquery', None)[
+                    'pageNo'
+                ] - 1
+            ) * common_job_parameters.get('pagination', {}).get('bigquey', None)['pageSize']
+            page_end = page_start + common_job_parameters.get('pagination', {}).get('bigquery', None)['pageSize']
+            if page_end > len(datasets) or page_end == len(datasets):
+                datasets = datasets[page_start:]
+            else:
+                has_next_page = True
+                datasets = datasets[page_start:page_end]
+                common_job_parameters['pagination']['bigquery']['hasNextPage'] = has_next_page
+
+        return datasets
+    except HttpError as e:
+        err = json.loads(e.content.decode('utf-8'))['error']
+        if err.get('status', '') == 'PERMISSION_DENIED' or err.get('message', '') == 'Forbidden':
+            logger.warning(
+                (
+                    "Could not retrieve Bigquery datasets on project %s due to permissions issues. Code: %s, Message: %s"
+                ), project_id, err['code'], err['message'],
+            )
+            return []
+        else:
+            raise
+
+
+@timeit
+def transform_bigquey_dataset(bigquery: Resource, datasets: List[Dict], project_id: str) -> List[Resource]:
+    list_dataset = []
+    for dataset in datasets:
+        dataset['id'] = dataset.get('datasetReference', {}).get('datasetId', '')
+        dataset['uniqueId'] = f"projects/{project_id}/datasets/{dataset['id']}"
+        dataset['details'] = get_dataset_info(bigquery, dataset['id'])
```

Comment:
we should populate consolelink property here

Reply:
The console link for dataset and tables are of the following format
`https://console.cloud.google.com/bigquery?referrer=search&project=cloudanix-qa-303120&ws=!1m4!1m3!3m2!1scloudanix-qa-303120!2sdataset1`. How should we implement the console link as ws field is different

Reply:
if we ignore the `ws` attribute does it fail?

Reply:
Added comment for future